symfit/core/minimizers.py

```diff
@@ -373,88 +435,62 @@ def scipy_constraints(self, constraints):
         """
         Returns all constraints in a scipy compatible format.
 
+        :param constraints: List of either MinimizeModel instances (this is what
+          is provided by :class:`~symfit.core.fit.Fit`),
+          :class:`~symfit.core.fit.Constraint`, or
+          :class:`sympy.core.relational.Relational`.
         :return: dict of scipy compatible statements.
         """
         cons = []
         types = {  # scipy only distinguishes two types of constraint.
             sympy.Eq: 'eq', sympy.Ge: 'ineq',
         }
 
-        for key, partialed_constraint in enumerate(constraints):
-            constraint_type = partialed_constraint.func.constraint_type
-            cons.append({
+        for constraint in constraints:
+            if isinstance(constraint, MinimizeModel):
+                # Typically the case when called by `Fit
+                constraint_type = constraint.model.constraint_type
+            elif hasattr(constraint, 'constraint_type'):
+                # Constraint object, not provided by `Fit`. Do the best we can.
+                if self.parameters != constraint.params:
+                    raise AssertionError('The constraint should accept the same'
+                                         ' parameters as used for the fit.')
+                constraint_type = constraint.constraint_type
+                constraint = MinimizeModel(constraint, data={})
+            elif isinstance(constraint, sympy.Rel):
+                from .fit import Constraint, CallableNumericalModel
+                # Todo: remove this CallableNumericalModel work around when
+                # either Constraint-obj are removed or a params arg is added.
+                constraint = Constraint(
+                    constraint,
+                    CallableNumericalModel({}, params=self.parameters,
+                                           independent_vars=[])
+                )
+                constraint_type = constraint.constraint_type
+                constraint = MinimizeModel(constraint, data={})
+            else:
+                raise TypeError('Unknown type for a constraint.')
+            con = {
                 'type': types[constraint_type],
-                # Takes an nd.array of params and a partialed_constraint, and
-                # evaluates the constraint with these parameters.
-                # Wrap `c` so it is always called via keywords.
-                'fun': lambda p, c: self.list2kwargs(c)(list(p))[0],
-                'args': [partialed_constraint]
-            })
-
+                'fun': constraint,
+                }
+            if isinstance(self, ScipyGradientMinimize):
+                con['jac'] = self.resize_jac(constraint.eval_jacobian)
```

Comment:
I'm not sure this is the right place for this. Maybe move it to ScipyGradientMinize (and ScipyHessianMinimize)?